### ISSUE TITLE
Improve vector quantization API

### ISSUE THREAD

**Issue by jpountz** [apache/lucene#11963]
### Description

Follow-up of https://github.com/apache/lucene/pull/11860#discussion_r1027106953: the API for quantization of vectors is a bit surprising at times in that you can index bytes but then still get float[] arrays whose values are bytes at search time. Can we make sure it's either bytes all the way, or floats all the way?

I'm not entirely sure how best to move it forward. One option would be to have a disjoint set of APIs for the binary encoding, like for doc values: different `Field` class, different `VectorValues` class, different `searchNearestNeighbors` method. I wonder if another option would be to make it floats all the way all the time, including for the byte encoding, ie. the `Field` and `VectorValues` class would still take and expose `float`s but `VectorValues#binaryValue` would be removed and the `Field` constructor would barf if the float values do not represent exact bytes.

**Comment by rmuir** [comment]
as i mentioned on the original issue(s) (there were several PRs closed and opened and many comments were lost...), the problem is right in the title: quantization.

there shouldnt be any quantization happening. otherwise it is broken.  8-bits in, 8-bits out, 32-bits in, 32-bits out. 

Additionally as a minimum there needs to be two separate field classes (e.g. ByteVectorValues/FloatVectorValues) to give type safety. Really, the types should be distinguished in fieldinfos today as well (not sure if that is currently the case).

 Basically just follow the examples of every other part of the index. vectors isn't special.

**Comment by rmuir** [comment]
we can break index backwards compatibility to fix this.

**Comment by jpountz** [comment]
> the types should be distinguished in fieldinfos today as well (not sure if that is currently the case)

They are, `FieldInfo#vectorEncoding`.

Trying to list what the change would look like:
 - Split `KnnVectorField` into `KnnFloatVectorField` and `KnnByteVectorField`.
 - Split `VectorValues` into `FloatVectorValues` and `ByteVectorValues`. `FloatVectorValues` would no longer expose its binary encoded value.
 - Split `VectorValues LeafReader#getVectorValues` into `ByteVectorValues LeafReader#getByteVectorValues` and `FloatVectorValues LeafReader#getFloatVectorValues`. These methods would return `null` if the field was indexed with the other type, like we're doing for doc values.
 - Split `LeafReader#searchNearestVectors` into one method that takes a `byte[]` and another one that takes a `float[]`. These methods would return empty hits if called with the wrong vector type, which would be consistent with exhaustive search by iterating over all vector values for the considered type.
 - Split `KnnVectorQuery` into `KnnFloatVectorQuery` and `KnnByteVectorQuery`

Some questions:
 - Returning `null` instances for the wrong type in `LeadReader#getFloatVectorValues` would be consistent with doc values, but I wonder if we should rather fail in order to catch issues early. I can't think of a single case when it would be desirable to be lenient?
 - Generics would require passing array classes to a number of these methods, which I don't like much for only two types that need to be supported?

**Comment by jtibshirani** [comment]
For context, we also have this issue suggesting refactors to the 8-bit quantization change: https://github.com/apache/lucene/issues/11758. I'm +1 on revising the API and hope it helps clean up some internal complexity too.

**Comment by benwtrent** [comment]
@jpountz 

> which I don't like much for only two types that need to be supported?

I do think we will add more types in the future. Specifically `binary` which can use an optimized hamming distance. This is useful for image search.

Splitting it all up like this is a big effort but I suppose its how it should have bene from the beginning.

Frustrating to block https://github.com/apache/lucene/pull/11860 for this. 

I will see about iterating on this refactor and see what the fallout is.

@rmuir if you have any further opinions on how you think this should look, let me know.

**Comment by rmuir** [comment]
@benwtrent go do your other issue first if you prefer. sorry for the trouble.

**Comment by benwtrent** [comment]
@rmuir thank you! I currently have a major part of the refactor already written locally. But, it will be big and will need its own rounds of review to make sure its the way we want it.

I want to try and get the PR up (once the size improvements is merged) ASAP

**Comment by benwtrent** [comment]
OK, step one of this is done with the `KnnByteVectorQuery`. I am next approaching a new `KnnByteVectorField` and this will cause some refactoring to `LeafReader` or `VectorValues`.

Right now, `VectorValues` assumes all vector values are `float[]` and "expands" the bytes, needlessly.

There are many ways to approach this and I am honestly not sure the way to go here (being unfamiliar with typical API patterns).

It seems there are a handful of seemingly valid options:

 * 1. Add a method to `VectorValues` (which is returned via `LeafReader`) that explicitly returns `byteVectorValue()` and will throw if `vectorValue()` is called when the encoding is `BYTE` (and throw on `byteVectorValue()` if encoding is `FLOAT32`).
    * This doesn't seem like the best to me. Though we already require users to choose the correct methods based on the vector encoding.
 * 2. Add a new `AbstractVectorValues<T>` where `vectorValue()` returns `T`. `LeafReader` would be changed to return that instead of `VectorValues`. 
    * There don't seem to be precedent on using generics like this anywhere, and this breaks with how we handle DocValues. The main difference with vectors and doc values is that numeric doc values are all returned as Long, but really could be any number whose bytes FIT in a long...
 * 3. Add a new `LeafReader#getByteVectorValues()` that returns a `ByteVectorValues` object. The main downside here is that we are not making it flexible for adding new vector kinds in the future. We will want to add boolean vectors & hamming distance in the future. These are important for image search.
 * 4. Simply make `VectorValues` always return `BytesRef` for everything and provide similarity that knows the encoding and can decode the `float[]` if that is the encoding. This feels closer to what we do with doc values (taking `DoubleDocValuesField` & `DoubleValuesSource` as prior art with how that interaction is done with `NumericDocValues`).


@jpountz @rmuir do y'all have opinions here?

**Comment by rmuir** [comment]
I would tend to expect some organization like this:

indexing:
* ByteVectorField
* FloatVectorField

searching:
* ByteVectorField.newQuery
* FloatVectorField.newQuery

codec api:
* KnnVectorsReader.getByteVectors(String field)
* KnnVectorsReader.getFloatVectors(String field)

**Comment by javanna** [comment]
It looks like we made all the suggested changes. One step that is missing is renaming the float classes and methods to follow exactly the naming convention suggested in the last comment. I opened #12105 for discussion.

**Comment by javanna** [comment]
#12105 and #12107 should complete the improvements that the description of this issue mentioned. I am leaning towards closing this if there are no objections.

**Comment by jpountz** [comment]
+1 to close
